### PR TITLE
fix(embedding): chunk EmbeddinggemmaONNX batches to bound ONNX memory (#1770)

### LINKED ISSUE #1770: repair --yes dies silently (kernel OOM): batch_size=5000 + EmbeddinggemmaONNX has no internal batching
## Summary

`mempalace repair --yes` dies **silently** (no traceback, no crash report, exit appears clean) on macOS ARM64 when the palace uses the `embeddinggemma` model. Root cause: `repair.py` hardcodes `batch_size = 5000` for the rebuild, and `EmbeddinggemmaONNX.__call__` has **no internal batching** — it tokenizes the entire input list and feeds one giant `session.run()` call. A 5000-document batch padded to the longest sequence allocates attention tensors far beyond available RAM; the kernel OOM-kills the process (SIGKILL → no `.ips` crash report, no Python traceback).

## Environment

- macOS 26.5 (Darwin 25.5) ARM64, Apple M5 Max, 128 GB RAM
- Python 3.12.13 (Homebrew), mempalace 3.4.0, chromadb 1.5.9, onnxruntime 1.26.0
- `config.json`: `embedding_model: embeddinggemma`, `embedding_device: cpu`
- Palace: ~9,000 drawers (originally built on mempalace 3.1.0 / chromadb 0.6.3, upgraded in place)

## Symptom

Every `mempalace repair --yes` run ends after the same line with no error:

```
  Building temporary collection: mempalace_drawers__repair_tmp
```

- Not a single `Staged N/total drawers...` line is ever printed (the first 5000-doc `upsert` never returns).
- Exit is silent: no `Repair failed:` message, no Python traceback, no DiagnosticReports `.ips` file (consistent with SIGKILL, not SIGSEGV/SIGBUS).
- Time to death scales with doc count (~1–6 min) — the time spent tokenizing + first allocations.
- Misleading UX: with output piped (`| tail`), the pipeline exit code is 0, so wrappers see "success".

Meanwhile `mine`, `search`, and the MCP server all work fine with embeddinggemma — they embed small batches only. Only repair's mega-batch dies.

## Root cause

1. `repair.py` (~line 792): `batch_size = 5000`, passed to `_rebuild_collection_via_temp`, so the first `temp_col.upsert()` carries 5000 documents.
2. `embedding.py` `EmbeddinggemmaONNX.__call__`: `encode_batch` on the full list → one `np.asarray` of shape `(5000, max_seq_len)` → **one** `self._session.run(...)` over the whole batch. No chunking. Transformer activation memory scales with `batch × seq_len²` per attention layer → tens-to-hundreds of GB for realistic drawer lengths → kernel OOM kill.

(For comparison, MiniLM survives 5000-doc batches because ChromaDB's built-in `ONNXMiniLM_L6_V2` chunks internally into batches of 32.)

## Repro

Any palace with a few thousand drawers + `embedding_model: embeddinggemma`:

```bash
mempalace repair --yes
# dies silently after "Building temporary collection: ..." — no Staged lines
```

## Workaround (verified)

Driving the library's own functions with a small rebuild batch completes normally:

```python
from mempalace.repair import (ChromaBackend, _drawers_collection_name,
                              _extract_drawers, _rebuild_collection_via_temp)
backend = ChromaBackend()
col = backend.get_collection(PALACE, _drawers_collection_name())
ids, docs, metas = _extract_drawers(col, col.count(), 5000)   # reads are fine
_rebuild_collection_via_temp(backend, PALACE, ids, docs, metas, 64)  # 64 works
```

`Staged 64/8937 drawers...` lines flow immediately and the rebuild proceeds to completion (8937/8937 re-filed, divergence 0).

## Suggested fix

1. Add internal chunking to `EmbeddinggemmaONNX.__call__` (e.g. 32–64 docs per `session.run`, mirroring ChromaDB's MiniLM EF) — this fixes every call site at once, and
2. Optionally lower repair's rebuild `batch_size` when the EF is embeddinggemma, and/or expose `--batch-size` on `repair`.
3. UX: flush a heartbeat/progress line *during* the first batch (related: #1486), so a wedged/killed embed is distinguishable from a slow one.

Related issues: #211 (ChromaDB batching at scale), #1486 (rebuild progress UX), #1355 (separate macOS 26 crash family — we hit that too during diagnosis, but this OOM reproduces independently of it).

### LINKED ISSUE #1784: CPU embedding (ONNX) exhausts system RAM on long mines — BFC arena grows unbounded, OOM kills desktop on 16 GB machines
## Environment
- MemPalace 3.4.0 (uv tool install), `embeddinggemma` (ONNX q8), backend chroma
- Windows 11, 16 GB RAM, mining ~9,000-file workspaces + Claude Code transcripts
- Follow-up to #1772 (same setup, different subsystem)

## Symptom
During large `mempalace mine` runs on CPU, the mining python process grows to **5.6–9.8 GB RSS** over hours. On a 16 GB machine this starved the OS twice: black screens, Explorer losing the desktop, other apps (Discord, the IDE host) failing to start, and finally:

```
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 :
RUNTIME_EXCEPTION : ... BFCArena::AllocateRawInternal Failed to allocate memory for
requested buffer of size 845991936
 in upsert.
```

```
Mine aborted by exception.
  files_processed: 147/268
  drawers_filed:   5966
```

## What it is NOT
- Not a single oversized file: the failing corpus had max line length < 1 KB and max file ~430 KB.
- Not fixed by `--max-chunks-per-file 300`: per-file caps don't bound the per-upsert embedding batch, and the ONNX **CPU BFC arena never shrinks** — it ratchets up across files until allocation fails or the OS suffocates. The 846 MB single-buffer request lands on top of an arena that had already grown to several GB.

## Workarounds that held up
1. **Bite-runner:** loop `mine --limit 200` so the process exits and frees its arena between bites. Crude but effective — RAM can never ratchet past one bite.
2. **The actual fix — GPU provider:** `mempalace[gpu]` + `embedding_device: cuda` moved the pressure into dedicated VRAM (~1.2 GB) and system RAM stayed at 7.9 GB free. Note: installing `mempalace[gpu]` left **both** `onnxruntime` and `onnxruntime-gpu` in the venv, and the CPU wheel shadowed the GPU one (`get_available_providers()` showed no CUDA). Had to `pip uninstall onnxruntime` + reinstall `onnxruntime-gpu` before `CUDAExecutionProvider` appeared. The extras resolution probably shouldn't allow both to coexist.

## Suggestions
- Cap the per-upsert embedding batch (e.g. embed in slices of 16–32 chunks) so a 200-chunk file can't demand an ~850 MB attention buffer on CPU.
- Consider `enable_cpu_mem_arena=False` (or periodic session re-creation) for the CPU provider on long mines — the arena ratchet is the killer, not any single file.
- Mining CLI could print RSS every N files; "slowly eats the machine over 6 hours" is hard to attribute without it.
- Fix the `[gpu]` extra so it replaces rather than coexists with the CPU `onnxruntime`.

Happy to provide logs or test a patch — same Windows test bench as #1772.

### PR DESCRIPTION
**Pull request by mvalentsev**
## What does this PR do?

Fixes #1770: `mempalace repair --yes` on an `embedding_model: embeddinggemma` palace dies silently (kernel OOM kill) after printing `Building temporary collection: ...`.

Closes #1784 as well. That report is the same unbounded ONNX batch on the mine path: long `mempalace mine` runs feed growing per-file batches into a single `session.run`, and the CPU arena's one-shot buffer request (846 MB there) eventually fails. The same 32-doc chunking bounds the embedder for both the repair path (#1770) and the mine path (#1784).

Root cause, confirmed on develop (2ec4bae): the repair paths upsert in batches of 5000 (`repair.py:792`, `cli.py:1038`), chromadb hands the whole list to the embedding function in one call (`CollectionCommon._embed`), and `EmbeddinggemmaONNX.__call__` tokenized the entire input and ran a single `session.run` over it. Attention buffers for a batch that size run to tens of GB even at short document lengths, growing superlinearly with document length, so the kernel kills the process before the first `Staged ...` line. ChromaDB's bundled MiniLM embedder avoids this because its `_forward` chunks internally at 32 docs per run; embeddinggemma had no equivalent.

The fix mirrors that pattern inside `EmbeddinggemmaONNX.__call__`: tokenize and run the model in sub-batches of 32 (constructor-tunable `batch_size`, validated >= 1), concatenating results in input order. This bounds peak memory at every call site at once: both 5000-doc repair paths, `rebuild_from_sqlite` (1000), the miners (`DRAWER_UPSERT_BATCH_SIZE = 1000` per-file batches), `migrate` (500), and oversized MCP `add_drawer`/diary writes. Empty input now returns `[]` instead of feeding a zero-row tensor to the runtime.

Measured (CPU, docs of ~190 tokens, address-space ceiling 2.3 GB as a stand-in for the reporter's kernel OOM kill):

| run | peak RSS | result |
|---|---|---|
| one-shot 256 docs (old code) | ceiling hit | dies: `BFCArena ... Failed to allocate memory` in `MultiHeadAttention` |
| sub-batches of 32, 256 docs (new code) | 1558 MB | completes |

One-shot growth is roughly linear in batch size (1001 MB embed-phase at 32 docs, 1433 MB at 128), consistent with the reporter's 128 GB machine dying on a 5000-doc batch of real drawers.

Vectors are unchanged: on 48 mixed-language docs of widely varying lengths, chunked vs single-batch embeddings differ by at most 1.2e-07 (min cosine 0.9999999); the model's pooled output is attention-masked, so per-chunk padding does not leak into values.

End-to-end on a real 600-drawer embeddinggemma palace under a 3.4 GB ceiling: develop's `repair --yes` dies inside the first upsert (`BFCArena` allocation failure in the first attention `MatMul`) with zero `Staged` lines, the reporter's exact symptom; with this change the same command stages and re-files 600/600 and completes.

Out of scope, by design: the repair-side `batch_size = 5000` upsert granularity (orthogonal once the embedding function is bounded; chromadb's own write cap is 5461) and progress output during the first batch (#1486 tracks that).

Thanks @mojie5 for the precise diagnosis; the fix follows the direction suggested in the issue.

## How to test

- `python -m pytest tests/test_embeddinggemma.py -v`: 18 tests, including chunk split and order pinning at 70 docs, boundary cases (1, 32, 33, 64), custom `batch_size`, rejection of `batch_size < 1`, and empty input.
- Manual: on a palace with a few thousand drawers and `embedding_model: embeddinggemma`, run `mempalace repair --yes`. Before: silent death after `Building temporary collection:`. After: `Staged N/total` lines flow and the rebuild completes.

## Checklist
- [x] Tests pass (`python -m pytest tests/ -v`)
- [x] No hardcoded paths
- [x] Linter passes (`ruff check .`)

### CODE REVIEW

File: mempalace/embedding.py [hunk 1]
```diff
@@ -158,10 +168,13 @@ def name() -> str:
...
         return "embeddinggemma_300m"
 
-    def __init__(self, preferred_providers=None):
+    def __init__(self, preferred_providers=None, batch_size: int = _EMBEDDINGGEMMA_BATCH_SIZE):
+        if batch_size < 1:
+            raise ValueError(f"batch_size must be >= 1, got {batch_size}")
         self._providers = (
             list(preferred_providers) if preferred_providers else ["CPUExecutionProvider"]
         )
+        self._batch_size = batch_size
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply mempalace/embedding.py:182]: Implemented in 4187521. `_lazy_load` now uses double-checked locking on an instance lock, with the session assigned last so the unlocked fast path can only observe a fully built loader. The same race existed one layer up: two threads missing `_EF_CACHE` would each keep a private instance and later load two models, so `get_embedding_function`'s check-then-construct is now also synchronized behind a lock-free fast path. Both paths are pinned by `test_concurrent_first_calls_load_model_once` and `test_concurrent_get_embedding_function_single_instance`.

File: mempalace/embedding.py [hunk 2]
```diff
@@ -210,21 +223,29 @@ def _lazy_load(self) -> None:
         self._tokenizer = tokenizer
         self._np = np
 
-    def __call__(self, input):  # noqa: A002 — ChromaDB EF protocol uses `input`
+    def __call__(self, input: list[str]) -> list[list[float]]:  # noqa: A002 — ChromaDB EF protocol
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply mempalace/embedding.py]: Done in 4187521. A bare str is wrapped as a single document, and `None`/empty inputs return `[]` before `_lazy_load` runs, so they no longer trigger the model download. The guard uses an explicit None-or-len check instead of truthiness so an array-like documents sequence is not rejected, and the `__call__` annotation now matches the accepted types. Covered by `test_call_bare_string_is_wrapped` and the extended `test_call_empty_input_returns_empty`.

**Comment by mvalentsev** [comment]
@gemini-code-assist both points are addressed in 4187521: double-checked locking on the lazy model load plus a lock on the EF cache's check-then-construct, and input guards in `__call__` (bare string wrapped, `None`/empty returned early before the model download). Details and test references are in the thread replies.
